File: scan_planner_ws/src/SCAN-Planner/src/planner/plan_env/src/near_field_cloud_fuser.cpp
#include <sensor_msgs/msg/point_cloud2.hpp>
#include <sensor_msgs/point_cloud2_iterator.hpp>

#include <rclcpp/rclcpp.hpp>

#include <algorithm>
#include <array>
#include <chrono>
#include <cmath>
#include <cstdint>
#include <deque>
#include <functional>
#include <limits>
#include <memory>
#include <stdexcept>
#include <string>
#include <unordered_set>
#include <vector>

namespace
{
using Point = std::array<float, 3>;
using Cloud = sensor_msgs::msg::PointCloud2;
using SteadyClock = std::chrono::steady_clock;

struct NearSample
{
  std::vector<Point> points;
  std::string frame;
  std::int64_t stamp_ns{0};
  SteadyClock::time_point received{};
  std::size_t candidates{0};
  std::size_t self_filtered{0};
  std::size_t height_filtered{0};
};

std::int64_t stampNanoseconds(const builtin_interfaces::msg::Time & stamp)
{
  return static_cast<std::int64_t>(stamp.sec) * 1000000000LL + stamp.nanosec;
}

bool finitePoint(float x, float y, float z)
{
  return std::isfinite(x) && std::isfinite(y) && std::isfinite(z);
}

std::uint64_t voxelKey(float x, float y, float z, double resolution)
{
  // Near-field coordinates are bounded to about +/-1.1 m. Sixteen bits per
  // signed voxel coordinate leave ample margin while producing a fast scalar
  // key for first-point-per-voxel filtering.
  const auto encode = [resolution](float value) {
    const auto cell = static_cast<std::int32_t>(std::floor(value / resolution));
    return static_cast<std::uint64_t>(
      static_cast<std::uint16_t>(cell + 32768));
  };
  return encode(x) | (encode(y) << 16U) | (encode(z) << 32U);
}
}  // namespace

class NearFieldCloudFuser : public rclcpp::Node
{
public:
  NearFieldCloudFuser()
  : Node("go2_near_field_cloud_fuser")
  {
    far_topic_ = declare_parameter<std::string>(
      "far_topic", "/lio_sam/deskew/cloud_deskewed");
    raw_topic_ = declare_parameter<std::string>(
      "raw_topic", "/unitree/slam_lidar/points");
    output_topic_ = declare_parameter<std::string>(
      "output_topic", "/scan_planner/local_cloud");
    near_min_ = declare_parameter<double>("near_min_range", 0.25);
    near_max_ = declare_parameter<double>("near_max_range", 1.05);
    near_voxel_ = declare_parameter<double>("near_voxel_size", 0.04);
    max_raw_age_ = declare_parameter<double>("max_raw_age", 0.25);
    max_raw_stamp_delta_ = declare_parameter<double>(
      "max_raw_stamp_delta", 0.03);

    base_to_lidar_x_ = declare_parameter<double>("base_to_lidar_x", 0.1701);
    base_to_lidar_y_ = declare_parameter<double>("base_to_lidar_y", 0.0);
    base_to_lidar_z_ = declare_parameter<double>("base_to_lidar_z", 0.0908);
    const double yaw = declare_parameter<double>(
      "base_to_lidar_yaw", 1.5707963267948966);
    mount_cos_ = std::cos(yaw);
    mount_sin_ = std::sin(yaw);

    body_radius_ = declare_parameter<double>("self_filter_radius", 0.24);
    body_offset_ = declare_parameter<double>("self_filter_offset", 0.12);
    self_z_min_ = declare_parameter<double>("self_filter_z_min", -0.55);
    self_z_max_ = declare_parameter<double>("self_filter_z_max", 0.30);
    obstacle_z_min_ = declare_parameter<double>("obstacle_body_z_min", -0.39);
    obstacle_z_max_ = declare_parameter<double>("obstacle_body_z_max", 0.55);

    if (!(near_min_ >= 0.0 && near_min_ < near_max_) || near_voxel_ <= 0.0) {
      throw std::invalid_argument("invalid near-field range or voxel size");
    }
    if (body_radius_ <= 0.0 || body_offset_ < 0.0) {
      throw std::invalid_argument("invalid self-filter footprint");
    }

    auto qos = rclcpp::SensorDataQoS().keep_last(1);
    publisher_ = create_publisher<Cloud>(output_topic_, qos);
    raw_subscription_ = create_subscription<Cloud>(
      raw_topic_, qos,
      std::bind(&NearFieldCloudFuser::onRaw, this, std::placeholders::_1));
    far_subscription_ = create_subscription<Cloud>(
      far_topic_, qos,
      std::bind(&NearFieldCloudFuser::onFar, this, std::placeholders::_1));
    last_log_ = SteadyClock::now() - std::chrono::seconds(5);

    RCLCPP_INFO(
      get_logger(),
      "Near-field fusion ready: %.2f..%.2f m raw XT16 + LIO deskew -> %s",
      near_min_, near_max_, output_topic_.c_str());
  }

private:
  void onRaw(const Cloud::ConstSharedPtr message)
  {
    std::vector<Point> filtered;
    std::unordered_set<std::uint64_t> occupied_voxels;
    filtered.reserve(1024);
    occupied_voxels.reserve(2048);
    std::size_t candidates = 0;
    std::size_t self_filtered = 0;
    std::size_t height_filtered = 0;
    const double near_min_squared = near_min_ * near_min_;
    const double near_max_squared = near_max_ * near_max_;
    const double body_radius_squared = body_radius_ * body_radius_;

    try {
      sensor_msgs::PointCloud2ConstIterator<float> x(*message, "x");
      sensor_msgs::PointCloud2ConstIterator<float> y(*message, "y");
      sensor_msgs::PointCloud2ConstIterator<float> z(*message, "z");
      for (; x != x.end(); ++x, ++y, ++z) {
        if (!finitePoint(*x, *y, *z)) {
          continue;
        }
        const double range_squared =
          static_cast<double>(*x) * *x + static_cast<double>(*y) * *y +
          static_cast<double>(*z) * *z;
        if (range_squared < near_min_squared || range_squared > near_max_squared) {
          continue;
        }
        ++candidates;

        // Lidar frame -> body frame, used only for filtering. Retained points
        // stay in the lidar frame consumed by SCAN's sensor_pose transform.
        const double body_x = mount_cos_ * *x - mount_sin_ * *y + base_to_lidar_x_;
        const double body_y = mount_sin_ * *x + mount_cos_ * *y + base_to_lidar_y_;
        const double body_z = *z + base_to_lidar_z_;
        const double front_distance_squared =
          (body_x - body_offset_) * (body_x - body_offset_) + body_y * body_y;
        const double rear_distance_squared =
          (body_x + body_offset_) * (body_x + body_offset_) + body_y * body_y;
        const bool inside_body =
          std::min(front_distance_squared, rear_distance_squared) <= body_radius_squared &&
          body_z >= self_z_min_ && body_z <= self_z_max_;
        if (inside_body) {
          ++self_filtered;
          continue;
        }
        if (body_z < obstacle_z_min_ || body_z > obstacle_z_max_) {
          ++height_filtered;
          continue;
        }

        const auto key = voxelKey(*x, *y, *z, near_voxel_);
        if (occupied_voxels.insert(key).second) {
          filtered.push_back(Point{*x, *y, *z});
        }
      }
    } catch (const std::runtime_error & error) {
      RCLCPP_WARN_THROTTLE(
        get_logger(), *get_clock(), 5000,
        "Raw cloud has no usable xyz fields: %s", error.what());
      filtered.clear();
    }

    NearSample sample;
    sample.points = std::move(filtered);
    sample.frame = message->header.frame_id;
    sample.stamp_ns = stampNanoseconds(message->header.stamp);
    sample.received = SteadyClock::now();
    sample.candidates = candidates;
    sample.self_filtered = self_filtered;
    sample.height_filtered = height_filtered;
    near_samples_.push_back(std::move(sample));
    // The matching deskewed cloud currently arrives about 0.2 s after the
    // raw scan. Keep enough timestamped scans to select the same acquisition
    // instead of accidentally fusing the next scan just because it arrived
    // most recently.
    while (near_samples_.size() > 16U) {
      near_samples_.pop_front();
    }
  }

  void onFar(const Cloud::ConstSharedPtr message)
  {
    const auto now = SteadyClock::now();
    const auto far_stamp_ns = stampNanoseconds(message->header.stamp);
    const NearSample * matching_near = nullptr;
    double matching_stamp_delta = std::numeric_limits<double>::infinity();
    for (const auto & sample : near_samples_) {
      if (sample.frame != message->header.frame_id) {
        continue;
      }
      const double delta = std::abs(
        static_cast<double>(far_stamp_ns - sample.stamp_ns) * 1e-9);
      if (delta < matching_stamp_delta) {
        matching_stamp_delta = delta;
        matching_near = &sample;
      }
    }
    const double matching_receive_age = matching_near ?
      std::chrono::duration<double>(now - matching_near->received).count() :
      std::numeric_limits<double>::infinity();
    const bool near_fresh = matching_near &&
      matching_receive_age <= max_raw_age_ &&
      matching_stamp_delta <= max_raw_stamp_delta_;
    const std::size_t near_count = near_fresh ? matching_near->points.size() : 0U;
    std::vector<Point> points;
    points.reserve(static_cast<std::size_t>(message->width) * message->height + near_count);
    std::size_t far_input_count = 0;
    std::size_t far_ground_filtered = 0;

    try {
      sensor_msgs::PointCloud2ConstIterator<float> x(*message, "x");
      sensor_msgs::PointCloud2ConstIterator<float> y(*message, "y");
      sensor_msgs::PointCloud2ConstIterator<float> z(*message, "z");
      for (; x != x.end(); ++x, ++y, ++z) {
        if (!finitePoint(*x, *y, *z)) {
          continue;
        }
        ++far_input_count;
        // Both inputs use the same lidar frame. Apply the lower body-height
        // cutoff to the deskewed cloud too, otherwise the 1 m handover leaves
        // concentric ground rings in the local obstacle layer.
        const double body_z = *z + base_to_lidar_z_;
        if (body_z < obstacle_z_min_) {
          ++far_ground_filtered;
          continue;
        }
        points.push_back(Point{*x, *y, *z});
      }
    } catch (const std::runtime_error & error) {
      RCLCPP_WARN_THROTTLE(
        get_logger(), *get_clock(), 5000,
        "Deskewed cloud has no usable xyz fields: %s", error.what());
      return;
    }
    const std::size_t far_count = points.size();
    if (near_fresh) {
      points.insert(
        points.end(), matching_near->points.begin(), matching_near->points.end());
    }

    Cloud output;
    output.header = message->header;
    output.height = 1;
    output.is_bigendian = false;
    output.is_dense = true;
    sensor_msgs::PointCloud2Modifier modifier(output);
    modifier.setPointCloud2FieldsByString(1, "xyz");
    modifier.resize(points.size());
    sensor_msgs::PointCloud2Iterator<float> output_x(output, "x");
    sensor_msgs::PointCloud2Iterator<float> output_y(output, "y");
    sensor_msgs::PointCloud2Iterator<float> output_z(output, "z");
    for (const auto & point : points) {
      *output_x = point[0];
      *output_y = point[1];
      *output_z = point[2];
      ++output_x;
      ++output_y;
      ++output_z;
    }
    publisher_->publish(output);

    if (std::chrono::duration<double>(now - last_log_).count() >= 5.0) {
      last_log_ = now;
      RCLCPP_INFO(
        get_logger(),
        "local_cloud: far=%zu/%zu ground=%zu near=%zu/%zu self=%zu height=%zu "
        "fresh=%s stamp_delta=%.3fs fused=%zu",
        far_count, far_input_count, far_ground_filtered, near_count,
        matching_near ? matching_near->candidates : 0U,
        matching_near ? matching_near->self_filtered : 0U,
        matching_near ? matching_near->height_filtered : 0U,
        near_fresh ? "true" : "false", matching_stamp_delta, points.size());
    }
  }

  std::string far_topic_;
  std::string raw_topic_;
  std::string output_topic_;
  double near_min_{0.25};
  double near_max_{1.05};
  double near_voxel_{0.04};
  double max_raw_age_{0.25};
  double max_raw_stamp_delta_{0.03};
  double base_to_lidar_x_{0.1701};
  double base_to_lidar_y_{0.0};
  double base_to_lidar_z_{0.0908};
  double mount_cos_{0.0};
  double mount_sin_{1.0};
  double body_radius_{0.24};
  double body_offset_{0.12};
  double self_z_min_{-0.55};
  double self_z_max_{0.30};
  double obstacle_z_min_{-0.39};
  double obstacle_z_max_{0.55};

  std::deque<NearSample> near_samples_;
  SteadyClock::time_point last_log_{};

  rclcpp::Publisher<Cloud>::SharedPtr publisher_;
  rclcpp::Subscription<Cloud>::SharedPtr raw_subscription_;
  rclcpp::Subscription<Cloud>::SharedPtr far_subscription_;
};

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<NearFieldCloudFuser>());
  rclcpp::shutdown();
  return 0;
}

File: scan_planner_ws/src/go2_scan_planner_bridge/go2_scan_planner_bridge/near_field_cloud_fuser.py
#!/usr/bin/env python3
"""Fuse LIO-SAM's deskewed cloud with a self-filtered XT16 near field.

LIO-SAM deliberately drops returns closer than ``lidarMinRange`` (currently
1.0 m).  That is reasonable for odometry but leaves a dangerous blind annulus
for local collision avoidance.  This node keeps the deskewed cloud unchanged
and adds only close raw XT16 returns after transforming them into the body
frame for height and self-body filtering.  The published points remain in the
lidar frame expected by SCAN's existing sensor-pose transform.
"""

import math
import time

import numpy as np
import rclpy
from rclpy.node import Node
from rclpy.qos import qos_profile_sensor_data
from sensor_msgs.msg import PointCloud2, PointField


XYZ_FIELDS = [
    PointField(name='x', offset=0, datatype=PointField.FLOAT32, count=1),
    PointField(name='y', offset=4, datatype=PointField.FLOAT32, count=1),
    PointField(name='z', offset=8, datatype=PointField.FLOAT32, count=1),
]


def cloud_xyz(message):
    """Return a compact float32 N×3 view/copy from an arbitrary PointCloud2."""
    offsets = {field.name: field.offset for field in message.fields}
    if not all(axis in offsets for axis in ('x', 'y', 'z')):
        return np.empty((0, 3), dtype=np.float32)
    endian = '>' if message.is_bigendian else '<'
    dtype = np.dtype({
        'names': ['x', 'y', 'z'],
        'formats': [endian + 'f4'] * 3,
        'offsets': [offsets['x'], offsets['y'], offsets['z']],
        'itemsize': int(message.point_step),
    })
    count = int(message.width * message.height)
    if count <= 0:
        return np.empty((0, 3), dtype=np.float32)
    cloud = np.frombuffer(message.data, dtype=dtype, count=count)
    points = np.column_stack((cloud['x'], cloud['y'], cloud['z'])).astype(
        np.float32, copy=False)
    return points[np.isfinite(points).all(axis=1)]


def voxel_sample(points, resolution):
    if len(points) == 0 or resolution <= 0.0:
        return points
    # np.unique(..., axis=0) creates a structured array and is surprisingly
    # expensive at lidar rate. Pack the three bounded voxel coordinates into
    # one int64 key so the same first-point policy uses the fast 1-D path.
    coordinates = np.floor(points / resolution).astype(np.int64)
    coordinates -= coordinates.min(axis=0)
    spans = coordinates.max(axis=0) + 1
    keys = ((coordinates[:, 0] * spans[1] + coordinates[:, 1]) *
            spans[2] + coordinates[:, 2])
    _, indices = np.unique(keys, return_index=True)
    return points[np.sort(indices)]


class NearFieldCloudFuser(Node):

    def __init__(self):
        super().__init__('go2_near_field_cloud_fuser')
        self.far_topic = str(self.declare_parameter(
            'far_topic', '/lio_sam/deskew/cloud_deskewed').value)
        self.raw_topic = str(self.declare_parameter(
            'raw_topic', '/unitree/slam_lidar/points').value)
        self.output_topic = str(self.declare_parameter(
            'output_topic', '/scan_planner/local_cloud').value)
        self.near_min = float(self.declare_parameter('near_min_range', 0.25).value)
        self.near_max = float(self.declare_parameter('near_max_range', 1.05).value)
        self.near_voxel = float(self.declare_parameter('near_voxel_size', 0.04).value)
        self.max_raw_age = float(self.declare_parameter('max_raw_age', 0.25).value)

        self.base_to_lidar_x = float(self.declare_parameter(
            'base_to_lidar_x', 0.1701).value)
        self.base_to_lidar_y = float(self.declare_parameter(
            'base_to_lidar_y', 0.0).value)
        self.base_to_lidar_z = float(self.declare_parameter(
            'base_to_lidar_z', 0.0908).value)
        yaw = float(self.declare_parameter(
            'base_to_lidar_yaw', math.pi / 2.0).value)
        self.mount_cos = math.cos(yaw)
        self.mount_sin = math.sin(yaw)

        self.body_radius = float(self.declare_parameter(
            'self_filter_radius', 0.24).value)
        self.body_offset = float(self.declare_parameter(
            'self_filter_offset', 0.12).value)
        self.self_z_min = float(self.declare_parameter(
            'self_filter_z_min', -0.55).value)
        self.self_z_max = float(self.declare_parameter(
            'self_filter_z_max', 0.30).value)
        self.obstacle_z_min = float(self.declare_parameter(
            'obstacle_body_z_min', -0.39).value)
        self.obstacle_z_max = float(self.declare_parameter(
            'obstacle_body_z_max', 0.55).value)

        if not (0.0 <= self.near_min < self.near_max):
            raise ValueError('near_min_range must be smaller than near_max_range')
        if self.body_radius <= 0.0 or self.body_offset < 0.0:
            raise ValueError('self-filter footprint parameters are invalid')

        self.raw_message = None
        self.raw_received_time = None
        self.last_processed_raw = None
        self.near_points = np.empty((0, 3), dtype=np.float32)
        self.near_frame = ''
        self.near_time = None
        self.raw_candidate_count = 0
        self.self_filtered_count = 0
        self.height_filtered_count = 0
        self.last_log = 0.0

        self.publisher = self.create_publisher(
            PointCloud2, self.output_topic, qos_profile_sensor_data)
        self.create_subscription(
            PointCloud2, self.raw_topic, self.on_raw, qos_profile_sensor_data)
        self.create_subscription(
            PointCloud2, self.far_topic, self.on_far, qos_profile_sensor_data)
        self.get_logger().info(
            'Near-field fusion ready: %.2f..%.2f m raw XT16 + LIO deskew -> %s' % (
                self.near_min, self.near_max, self.output_topic))

    def on_raw(self, message):
        # The fused output is clocked by the deskewed LIO cloud. Keep only the
        # newest raw frame here, then filter it once when the next far frame
        # arrives. Processing every intermediate raw frame wastes CPU without
        # producing an additional fused observation.
        self.raw_message = message
        self.raw_received_time = time.monotonic()

    def update_near(self, message, received_time):
        points = cloud_xyz(message)
        if len(points) == 0:
            self.near_points = np.empty((0, 3), dtype=np.float32)
            self.near_frame = message.header.frame_id
            self.near_time = received_time
            return

        ranges = np.linalg.norm(points, axis=1)
        candidate_mask = (ranges >= self.near_min) & (ranges <= self.near_max)
        candidates = points[candidate_mask]
        self.raw_candidate_count = int(len(candidates))
        if len(candidates) == 0:
            self.near_points = np.empty((0, 3), dtype=np.float32)
            self.near_frame = message.header.frame_id
            self.near_time = received_time
            return

        # Lidar frame -> body frame, used only to decide what to retain.
        body_x = (self.mount_cos * candidates[:, 0] -
                  self.mount_sin * candidates[:, 1] + self.base_to_lidar_x)
        body_y = (self.mount_sin * candidates[:, 0] +
                  self.mount_cos * candidates[:, 1] + self.base_to_lidar_y)
        body_z = candidates[:, 2] + self.base_to_lidar_z

        front_distance2 = (body_x - self.body_offset) ** 2 + body_y ** 2
        rear_distance2 = (body_x + self.body_offset) ** 2 + body_y ** 2
        within_capsule = np.minimum(front_distance2, rear_distance2) <= (
            self.body_radius ** 2)
        within_self_height = ((body_z >= self.self_z_min) &
                              (body_z <= self.self_z_max))
        self_mask = within_capsule & within_self_height
        height_mask = ((body_z >= self.obstacle_z_min) &
                       (body_z <= self.obstacle_z_max))
        keep = height_mask & ~self_mask
        self.self_filtered_count = int(self_mask.sum())
        self.height_filtered_count = int((~height_mask & ~self_mask).sum())
        self.near_points = voxel_sample(candidates[keep], self.near_voxel)
        self.near_frame = message.header.frame_id
        self.near_time = received_time

    def on_far(self, message):
        far_points = cloud_xyz(message)
        if len(far_points) == 0:
            return
        far_input_count = len(far_points)
        far_keep = ((far_points[:, 2] + self.base_to_lidar_z) >=
                    self.obstacle_z_min)
        far_ground_filtered = int((~far_keep).sum())
        far_points = far_points[far_keep]
        now = time.monotonic()
        if (self.raw_message is not None and
                self.raw_message is not self.last_processed_raw):
            raw_message = self.raw_message
            raw_received_time = self.raw_received_time
            self.update_near(raw_message, raw_received_time)
            self.last_processed_raw = raw_message
        near_fresh = (self.near_time is not None and
                      now - self.near_time <= self.max_raw_age and
                      self.near_frame == message.header.frame_id)
        if near_fresh and len(self.near_points):
            fused = np.concatenate((far_points, self.near_points), axis=0)
        else:
            fused = far_points

        fused = np.ascontiguousarray(fused, dtype='<f4')
        output = PointCloud2()
        output.header = message.header
        output.height = 1
        output.width = int(len(fused))
        output.fields = XYZ_FIELDS
        output.is_bigendian = False
        output.point_step = 12
        output.row_step = output.point_step * output.width
        output.is_dense = bool(np.isfinite(fused).all())
        output.data = fused.tobytes()
        self.publisher.publish(output)

        if now - self.last_log >= 5.0:
            self.last_log = now
            self.get_logger().info(
                'local_cloud: far=%d/%d ground=%d near=%d/%d self=%d height=%d fresh=%s fused=%d' % (
                    len(far_points), far_input_count, far_ground_filtered,
                    len(self.near_points), self.raw_candidate_count,
                    self.self_filtered_count, self.height_filtered_count,
                    near_fresh, len(fused)))


def main(args=None):
    rclpy.init(args=args)
    node = NearFieldCloudFuser()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == '__main__':
    main()
